grid2op/Reward/_alertTrustScore.py

```diff
@@ -0,0 +1,163 @@
+                 reward_max_blackout=0.0,
+                 reward_end_episode_bonus=0.0,
+                 min_score=-3):
+
+        super().__init__(logger,
+                 reward_min_no_blackout,
+                 reward_min_blackout, 
+                 reward_max_no_blackout,
+                 reward_max_blackout,
+                 reward_end_episode_bonus)
+        
+        self.min_score = dt_float(min_score)
+        self.max_score = dt_float(1.0)
+        self.blackout_encountered = False
+        
+    def initialize(self, env):
+        self._is_simul_env = self.is_simulated_env(env)
+            
+        self.cumulated_reward = 0
+        #KPIs
+        self.total_nb_attacks = 0
+        self.nb_last_attacks = 0 #attacks in the AlertTimeWindow before done
+
+        #TODO
+        #self.total_nb_alerts = 0
+        #self.alert_attack_no_blackout = 0
+        #self.alert_attack_blackout = 0
+        #self.no_alert_attack_no_blackout = 0
+        #self.no_alert_attack_blackout = 0
+        #self.blackout_encountered = False
+        return super().initialize(env)
+            
+    def reset(self, env):
+        super().reset(env)
+        self.cumulated_reward = 0
+        #KPIs
+        self.total_nb_attacks = 0
+        self.nb_last_attacks = 0
+
+        # TODO
+        #self.total_nb_alerts = 0
+        #self.alert_attack_no_blackout = 0
+        #self.alert_attack_blackout = 0
+        #self.no_alert_attack_no_blackout = 0
+        #self.no_alert_attack_blackout = 0
+        #self.blackout_encountered = False
+            
+    def __call__(self, action, env, has_error, is_done, is_illegal, is_ambiguous):
+
+        score_ep = 0.
+        if self._is_simul_env:
+            return score_ep
+
+        self.blackout_encountered = self.is_in_blackout(has_error, is_done)
+        
+        score_ep = 0.
+        if self._is_simul_env:
+            return score_ep
+            
+        res = super().__call__(action, env, has_error, is_done, is_illegal, is_ambiguous)
+        self.cumulated_reward += res
+
+        is_attack = (env._time_since_last_attack == 0).any()#even if there are simultaneous attacks, we consider this as a single attack event
+        self.total_nb_attacks += is_attack
+
+        # TODO
+        #lines_alerted_beforeattack = np.equal(env._time_since_last_alert, env._time_since_last_attack + 1) and lines_attacked
+        #self.total_nb_alerts += np.sum(lines_alerted_beforeattack)
+            
+        if not is_done:
+
+            # TODO
+            #lines_attacked_no_blackout = env._time_since_last_attack == SURVIVOR_TIMESTEPS
+#
+            #self.alert_attack_no_blackout += np.sum(lines_alerted_beforeattack[lines_attacked_no_blackout])
+            #self.no_alert_attack_no_blackout += np.sum(~lines_alerted_beforeattack[lines_attacked_no_blackout])
+            
+            return score_ep
+            
+        else:
+            self.nb_last_attacks=np.sum(np.any(self._ts_attack,axis=1))
```

Comment:
Ok after a deep look it appears to be working because of the implementation of AlertReward which put to "False" the _ts_attack right after it has been processed by this class.

So maybe the bug is not where I thought it would be but I wonder now if self.nb_last_attacks can be 1  sometimes.

Reply:
And it does not "break" because _ts_attack is NOT reset to `False`  in the base class when there is a game over.

So it's a real miracle that heavily depends on the behaviour of the implementation of the base class. This is not recommended.

At least use:

```python
            indexes_to_look = (np.arange(-self.time_window, 1) + self._current_id) % self._nrows_array  # include current step (hence the np.arange(..., **1**))
            ts_attack_in_order = self._ts_attack[indexes_to_look, :]
            self.nb_last_attacks = ts_attack_in_order.any(axis=1).sum()
```

So that it at least does not depend on the implementation of the base class

Reply:
A even better version would include a function (in AlertReward) that does the computation of `ts_attack_in_order` so that nothing is copy pasted.

Reply:
ok I added the code block for now